### ISSUE TITLE
Different SPARQL engines treat `GROUP_CONCAT` differently

### ISSUE THREAD

**Issue by AlasdairGray** [AlasdairGray/IDP-KG#19]
The rdflib SPARQL engine permits none `GROUP BY` variables to be returned in the `SELECT` clause while GraphDB and Fuseki complain.

RDFlib engine does not deal with no values being returned to the `GROUP_CONCAT` operator and returns an error.

This issue was originally noted in [comment on #17](https://github.com/AlasdairGray/IDPcentral/issues/17#issuecomment-909020460)
> Note that the Full Protein Information query will not run on a standards compliant endpoint. It returns variables that are not in the `GROUP BY` clause. This is because RDFlib returns an error if there are no bindings for a `GROUP_CONCAT`. The correct query should be
> 
> ```sparql
> PREFIX schema: <https://schema.org/>
> PREFIX pav: <http://purl.org/pav/>
> PREFIX void: <http://rdfs.org/ns/void#>
> 
> SELECT  ?s 
>     (GROUP_CONCAT(DISTINCT ?name;SEPARATOR=',<br/>') AS ?names)
>     (GROUP_CONCAT(DISTINCT ?description;SEPARATOR=',<br/>') AS ?descriptions)
>     (GROUP_CONCAT(DISTINCT ?identifier;SEPARATOR=',<br/>') AS ?identifiers)
>     (GROUP_CONCAT(DISTINCT ?associatedDisease;SEPARATOR=',<br/>') AS ?diseases)
>     (GROUP_CONCAT(DISTINCT ?annotation;SEPARATOR=',<br/>') AS ?annotations)
>     (GROUP_CONCAT(DISTINCT ?encodedBy;SEPARATOR=',<br/>') AS ?genes)
>     (GROUP_CONCAT(DISTINCT ?taxonomicRange;SEPARATOR=',<br/>') AS ?taxons)
>     (GROUP_CONCAT(DISTINCT ?url;SEPARATOR=',<br/>') AS ?urls)
>     (GROUP_CONCAT(DISTINCT ?alternateName;SEPARATOR=',<br/>') AS ?altNames)
>     (GROUP_CONCAT(DISTINCT ?bioChemInteraction;SEPARATOR=',<br/>') AS ?interactions)
>     (GROUP_CONCAT(DISTINCT ?bioChemSimilarity;SEPARATOR=',<br/>') AS ?similarTo)
>     (GROUP_CONCAT(DISTINCT ?bioChemEntity;SEPARATOR=',<br/>') AS ?entities)
>     (GROUP_CONCAT(DISTINCT ?sequence;SEPARATOR=',<br/>') AS ?sequences)
>     (GROUP_CONCAT(DISTINCT ?molFunction;SEPARATOR=',<br/>') AS ?functions)
>     (GROUP_CONCAT(DISTINCT ?representation;SEPARATOR=',<br/>') AS ?representations)
>     (GROUP_CONCAT(DISTINCT ?image;SEPARATOR=',<br/>') AS ?images)
>     (GROUP_CONCAT(DISTINCT ?process;SEPARATOR=',<br/>') AS ?processes)
>     (GROUP_CONCAT(DISTINCT ?cellularLocation;SEPARATOR=',<br/>') AS ?locations)
>     (GROUP_CONCAT(DISTINCT ?parentEntity;SEPARATOR=',<br/>') AS ?parents)
>     (GROUP_CONCAT(DISTINCT ?sameAs;SEPARATOR=',<br/>') AS ?sameAs_s)
>     (GROUP_CONCAT(DISTINCT ?source;SEPARATOR=',<br/>') AS ?sources)
>     (GROUP_CONCAT(DISTINCT ?dataset;SEPARATOR=',<br/>') AS ?datasets)
> WHERE {
>     GRAPH ?g {
> # Bioschemas Minimal Properties
>         ?s a schema:Protein .
>         OPTIONAL {?s schema:identifier ?identifier }
>         OPTIONAL {?s schema:name ?name }
> ## Bioschemas Recommended properties
>         OPTIONAL {?s schema:associatedDisease ?associatedDisease}
>         OPTIONAL {?s schema:description ?description}
>         #OPTIONAL 
>         {?s schema:hasSequenceAnnotation ?annotation }
>         OPTIONAL {?s schema:isEncodedByBioChemEntity ?encodedBy}
>         OPTIONAL {?s schema:taxonomicRange ?taxonomicRange }
>         OPTIONAL {?s schema:url ?url}
> ## Bioschemas Optional properties
>         OPTIONAL {?s schema:alternateName ?alternateName}
>         OPTIONAL {?s schema:bioChemInteraction ?bioChemInteraction}
>         OPTIONAL {?s schema:bioChemSimilarity ?bioChemSimilarity}
>         OPTIONAL {?s schema:hasBioChemEntityPart ?bioChemEntity}
>         OPTIONAL {?s schema:hasBioPolymerSequence ?sequence}
>         OPTIONAL {?s schema:hasMolecularFunction ?molFunction}
>         OPTIONAL {?s schema:hasRepresentation ?representation }
>         OPTIONAL {?s schema:image ?image}
>         OPTIONAL {?s schema:isInvolvedInBiologicalProcess ?process}
>         OPTIONAL {?s schema:isLocatedInSubcellularLocation ?cellularLocation}
>         OPTIONAL {?s schema:isPartOfBioChemEntity ?parentEntity}
>         OPTIONAL {?s schema:sameAs ?sameAs }
>     }
>     ?g pav:retrievedFrom ?source ;
>     OPTIONAL {?g void:inDataset ?dataset}
> }
> GROUP BY ?s
> ```

**Comment by AlasdairGray** [comment]
Behaviour of GraphDB and Fuseki is correct. RDFlib does not have the correct implementation of `GROUP_CONCAT`.

**Comment by AlasdairGray** [comment]
@ImranAsif48 will investigate the possibility of using [pyHDT](https://github.com/Callidon/pyHDT) which is included as part of RDFlib 6.0. Hopefully this will allow us to provide an in notebook query engine that is SPARQL compliant.

**Comment by ImranAsif48** [comment]
@AlasdairGray I tried to install the rdflib-hdt using pip3 install command. But it is giving me the error. I also installed the gcc/clang with c++11 support before installing this package. **(Note: Very time consuming work)**

After that I found the solution to generate the hdt file from rdf file using command line and we can also run the sparql query as command line. I am using Python to run these commands on jupyter notebook. The issue is that we can show the results in very bad format due to command line output.

The rdf2hdt conversion is very fast. I gave IDPKG_Full.nq (33.2MB) and it took only 39ms
